File: internal/api/handlers/projects.go
package handlers

import (
	"chklst-go/internal/database"
	"fmt"
	"strconv"

	"github.com/gofiber/fiber/v3"
	"gorm.io/gorm"
)

// uniqueProjectName returns a project name that does not yet exist, appending
// " (copy)", " (copy 2)", ... to the desired base name until a free slot is found.
func uniqueProjectName(tx *gorm.DB, base string) string {
	candidate := base
	for i := 1; ; i++ {
		var count int64
		tx.Model(&database.Project{}).Where("name = ?", candidate).Count(&count)
		if count == 0 {
			return candidate
		}
		if i == 1 {
			candidate = base + " (copy)"
		} else {
			candidate = fmt.Sprintf("%s (copy %d)", base, i)
		}
	}
}

// ListProjects returns all projects
func ListProjects(c fiber.Ctx) error {
	var projects []database.Project
	
	// Preload components
	if err := database.DB.Preload("Components").Find(&projects).Error; err != nil {
		return c.Status(500).JSON(fiber.Map{
			"error": "Failed to fetch projects",
		})
	}

	return c.JSON(projects)
}

// GetProject returns a single project by ID
func GetProject(c fiber.Ctx) error {
	id, err := strconv.Atoi(c.Params("id"))
	if err != nil {
		return c.Status(400).JSON(fiber.Map{
			"error": "Invalid project ID",
		})
	}

	var project database.Project
	if err := database.DB.Preload("Components").First(&project, id).Error; err != nil {
		return c.Status(404).JSON(fiber.Map{
			"error": "Project not found",
		})
	}

	return c.JSON(project)
}

// CreateProject creates a new project
func CreateProject(c fiber.Ctx) error {
	var project database.Project

	if err := c.Bind().JSON(&project); err != nil {
		return c.Status(400).JSON(fiber.Map{
			"error": "Invalid request body",
		})
	}

	// Validate required fields
	if project.Name == "" {
		return c.Status(400).JSON(fiber.Map{
			"error": "Project name is required",
		})
	}

	if err := database.DB.Create(&project).Error; err != nil {
		return c.Status(500).JSON(fiber.Map{
			"error": "Failed to create project",
		})
	}

	return c.Status(201).JSON(project)
}

// UpdateProject updates an existing project
func UpdateProject(c fiber.Ctx) error {
	id, err := strconv.Atoi(c.Params("id"))
	if err != nil {
		return c.Status(400).JSON(fiber.Map{
			"error": "Invalid project ID",
		})
	}

	var project database.Project
	if err := database.DB.First(&project, id).Error; err != nil {
		return c.Status(404).JSON(fiber.Map{
			"error": "Project not found",
		})
	}

	if err := c.Bind().JSON(&project); err != nil {
		return c.Status(400).JSON(fiber.Map{
			"error": "Invalid request body",
		})
	}

	if err := database.DB.Save(&project).Error; err != nil {
		return c.Status(500).JSON(fiber.Map{
			"error": "Failed to update project",
		})
	}

	// Reload with components so the response carries the full project. Otherwise
	// the client would replace its state with a component-less project and the
	// component list would appear to vanish until the next refresh.
	database.DB.Preload("Components").First(&project, id)
	return c.JSON(project)
}

// DuplicateProject deep-clones a project (and all its components) under a new name.
// Body: { "new_name": "..." } (optional; defaults to "<name> (copy)").
// The clone is independent — deployments are NOT copied.
func DuplicateProject(c fiber.Ctx) error {
	id, err := strconv.Atoi(c.Params("id"))
	if err != nil {
		return c.Status(400).JSON(fiber.Map{
			"error": "Invalid project ID",
		})
	}

	var source database.Project
	if err := database.DB.Preload("Components").First(&source, id).Error; err != nil {
		return c.Status(404).JSON(fiber.Map{
			"error": "Project not found",
		})
	}

	// Optional new name from body
	var body struct {
		NewName string `json:"new_name"`
	}
	_ = c.Bind().JSON(&body) // body is optional; ignore bind errors

	base := body.NewName
	if base == "" {
		base = source.Name + " (copy)"
	}

	var clone database.Project
	err = database.DB.Transaction(func(tx *gorm.DB) error {
		clone = database.Project{
			Name:           uniqueProjectName(tx, base),
			BuildServer:    source.BuildServer,
			DeployServer:   source.DeployServer,
			DatabaseName:   source.DatabaseName,
			Environment:    source.Environment,
			BackupLocation: source.BackupLocation,
			Description:    source.Description,
		}
		if err := tx.Create(&clone).Error; err != nil {
			return err
		}

		for _, comp := range source.Components {
			newComp := database.Component{
				ProjectID:    clone.ID,
				Name:         comp.Name,
				Developer:    comp.Developer,
				VCSType:      comp.VCSType,
				VCSURL:       comp.VCSURL,
				BuildCommand: comp.BuildCommand,
				ComponentURL: comp.ComponentURL,
				Enabled:      comp.Enabled,
				Description:  comp.Description,
			}
			if err := tx.Create(&newComp).Error; err != nil {
				return err
			}
		}
		return nil
	})
	if err != nil {
		return c.Status(500).JSON(fiber.Map{
			"error": "Failed to duplicate project",
		})
	}

	// Reload with components for the response
	database.DB.Preload("Components").First(&clone, clone.ID)
	return c.Status(201).JSON(clone)
}

// DeleteProject deletes a project
func DeleteProject(c fiber.Ctx) error {
	id, err := strconv.Atoi(c.Params("id"))
	if err != nil {
		return c.Status(400).JSON(fiber.Map{
			"error": "Invalid project ID",
		})
	}

	if err := database.DB.Delete(&database.Project{}, id).Error; err != nil {
		return c.Status(500).JSON(fiber.Map{
			"error": "Failed to delete project",
		})
	}

	return c.SendStatus(204)
}

File: frontend/src/stores/projects.ts
import { defineStore } from 'pinia'
import { ref, computed } from 'vue'
import { useApi } from '../composables/useApi'

export interface Component {
  id?: number
  name: string
  developer?: string
  vcs_type?: 'git' | 'svn'
  vcs_url?: string
  build_command?: string
  component_url?: string
  enabled?: boolean
  description?: string
  created_at?: string
  updated_at?: string
}

export interface Project {
  id?: number
  name: string
  build_server?: string
  deploy_server?: string
  database_name?: string
  environment?: string
  backup_location?: string
  description?: string
  components: Component[]
  created_at?: string
  updated_at?: string
}

export const useProjectsStore = defineStore('projects', () => {
  const projects = ref<Project[]>([])
  const isLoading = ref(false)
  const error = ref<string | null>(null)
  const { get, post, put, delete: deleteApi } = useApi()

  const fetchProjects = async () => {
    isLoading.value = true
    error.value = null
    try {
      const response = await get<Project[]>('/projects')
      // Sort projects alphabetically by name (ascending)
      const sortedProjects = (response.data || []).sort((a, b) =>
        (a.name || '').localeCompare(b.name || '')
      )
      projects.value = sortedProjects
    } catch (err) {
      error.value = 'Failed to fetch projects'
      console.error(err)
    } finally {
      isLoading.value = false
    }
  }

  const createProject = async (project: Project): Promise<Project | null> => {
    isLoading.value = true
    error.value = null
    try {
      const response = await post<Project>('/projects', project)
      const newProject = response.data
      if (newProject.id) {
        projects.value.push(newProject)
      }
      return newProject
    } catch (err) {
      error.value = 'Failed to create project'
      console.error(err)
      return null
    } finally {
      isLoading.value = false
    }
  }

  const updateProject = async (id: number, project: Partial<Project>) => {
    isLoading.value = true
    error.value = null
    try {
      const response = await put<Project>(`/projects/${id}`, project)
      const idx = projects.value.findIndex(p => p.id === id)
      if (idx > -1) {
        projects.value[idx] = response.data
      }
      return response.data
    } catch (err) {
      error.value = 'Failed to update project'
      console.error(err)
      return null
    } finally {
      isLoading.value = false
    }
  }

  const deleteProject = async (id: number) => {
    isLoading.value = true
    error.value = null
    try {
      await deleteApi(`/projects/${id}`)
      projects.value = projects.value.filter(p => p.id !== id)
    } catch (err) {
      error.value = 'Failed to delete project'
      console.error(err)
    } finally {
      isLoading.value = false
    }
  }

  const addComponent = async (projectId: number, component: Component) => {
    isLoading.value = true
    error.value = null
    try {
      const response = await post<Component>(
        `/projects/${projectId}/components`,
        component
      )
      const project = projects.value.find(p => p.id === projectId)
      if (project) {
        project.components.push(response.data)
      }
      return response.data
    } catch (err) {
      error.value = 'Failed to add component'
      console.error(err)
      return null
    } finally {
      isLoading.value = false
    }
  }

  const updateComponent = async (
    projectId: number,
    componentId: number,
    component: Partial<Component>
  ) => {
    isLoading.value = true
    error.value = null
    try {
      const response = await put<Component>(
        `/projects/${projectId}/components/${componentId}`,
        component
      )
      const project = projects.value.find(p => p.id === projectId)
      if (project) {
        const idx = project.components.findIndex(c => c.id === componentId)
        if (idx > -1) {
          project.components[idx] = response.data
        }
      }
      return response.data
    } catch (err) {
      error.value = 'Failed to update component'
      console.error(err)
      return null
    } finally {
      isLoading.value = false
    }
  }

  const deleteComponent = async (projectId: number, componentId: number) => {
    isLoading.value = true
    error.value = null
    try {
      await deleteApi(`/projects/${projectId}/components/${componentId}`)
      const project = projects.value.find(p => p.id === projectId)
      if (project) {
        project.components = project.components.filter(c => c.id !== componentId)
      }
    } catch (err) {
      error.value = 'Failed to delete component'
      console.error(err)
    } finally {
      isLoading.value = false
    }
  }

  const getProjectByName = (name: string) => {
    return projects.value.find(p => p.name === name)
  }

  const getComponentsByProject = (projectId: number) => {
    const project = projects.value.find(p => p.id === projectId)
    return project?.components || []
  }

  // Combined Project+Component options for single dropdown
  const projectComponentOptions = computed(() => {
    const options: { value: string; label: string; projectId: number; componentId: number; project: Project; component: Component }[] = []
    for (const project of projects.value) {
      for (const component of project.components || []) {
        options.push({
          value: `${project.id}-${component.id}`,
          label: `${project.name} > ${component.name}`,
          projectId: project.id!,
          componentId: component.id!,
          project,
          component,
        })
      }
    }
    return options
  })

  return {
    projects,
    isLoading,
    error,
    projectComponentOptions,
    fetchProjects,
    createProject,
    updateProject,
    deleteProject,
    addComponent,
    updateComponent,
    deleteComponent,
    getProjectByName,
    getComponentsByProject,
  }
})
